File: frontend/iosApp/Domains/Sources/Domains/OptionalUpdate.swift
import Foundation

/// Represents an optional update operation for a field.
/// Used in `with` copy methods to distinguish between:
/// - Not changing the field
/// - Explicitly setting the field to nil
/// - Setting a new value
public enum OptionalUpdate<T> {
    /// Keep the current value (don't update)
    case noChange
    /// Set to nil
    case setNil
    /// Set to a new value
    case set(T)

    /// Creates an OptionalUpdate from an optional value
    /// - nil becomes .setNil
    /// - value becomes .set(value)
    public static func from(_ value: T?) -> OptionalUpdate<T> {
        if let value = value {
            return .set(value)
        } else {
            return .setNil
        }
    }

    /// Resolves the update against the current value
    public func resolve(current: T?) -> T? {
        switch self {
        case .noChange:
            return current
        case .setNil:
            return nil
        case .set(let value):
            return value
        }
    }
}

File: frontend/shared/src/commonMain/kotlin/com/example/memoriesapp/domain/OptionalUpdate.kt
package com.example.memoriesapp.domain

/**
 * Represents an optional update operation for a field.
 * Used in `copy` methods to distinguish between:
 * - Not changing the field
 * - Explicitly setting the field to null
 * - Setting a new value
 */
sealed class OptionalUpdate<out T> {
    /** Keep the current value (don't update) */
    data object NoChange : OptionalUpdate<Nothing>()

    /** Set to null */
    data object SetNull : OptionalUpdate<Nothing>()

    /** Set to a new value */
    data class Set<T>(val value: T) : OptionalUpdate<T>()

    companion object {
        /**
         * Creates an OptionalUpdate from a nullable value
         * - null becomes SetNull
         * - value becomes Set(value)
         */
        fun <T> from(value: T?): OptionalUpdate<T> =
            if (value != null) Set(value) else SetNull
    }

    /**
     * Resolves the update against the current value
     */
    fun resolve(current: @UnsafeVariance T?): T? = when (this) {
        is NoChange -> current
        is SetNull -> null
        is Set -> value
    }
}
